092-self-healing-using-cruise-control.md

```diff
@@ -0,0 +1,273 @@
+```
+the detected anomaly would look something like this:
+```shell
+#...
+recentDiskFailures:[{anomalyId=4b1ed191-50d7-4aa1-806b-b8c5c0d679fe, 
+detectionDate=2024-12-09T09:49:46Z, 
+statusUpdateDate=2024-12-09T09:49:47Z, 
+failedDisksByTimeMs={2={/var/lib/kafka/data-1/kafka-log2=1733737786603}}, 
+status=FIX_STARTED}]}
+```
+where, 
+* `anomalyId` represents the task id assigned to the anomaly
+* `failedDisksByTimeMs` represents the broker id(which in this case is 2) as well the disk(/var/lib/kafka/data-1/kafka-log2) which failed, 
+* `detectionDate` and `statusUpdateDate` represent the date and time when the anomaly was detected and the status of the anomaly was updated.
+
+The anomaly status would change based on how the healing progresses. The anomaly can transition to these possible status:
+* DETECTED - The anomaly is just detected and no action is yet taken on it 
+* IGNORED - Self healing is either disabled or the anomaly is unfixable
+* FIX_STARTED - The anomaly fix has started
+* FIX_FAILED_TO_START - The proposal generation for the anomaly fix failed 
+* CHECK_WITH_DELAY - The anomaly fix is delayed 
+* LOAD_MONITOR_NOT_READY - Monitor for workload of a Kafka cluster is in loading or bootstrap state
+* COMPLETENESS_NOT_READY - Completeness not met for some goals. The completeness describes the confidence level of the data in the metric sample aggregator. If this is low then some Goal classes may refuse to produce proposals.
+
+We can again query the [`state` endpoint](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control), with `substate` parameter as `executor`, to get details on the tasks that are currently being performed by Cruise Control.
+For example:
+```shell
+ExecutorState: {state: INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS,
+ pending(20)/in-progress(0)/aborting(0)/finished(0)/total(20) inter-broker partition movements, 
+ completed(0)/total(0) bytes in MBs: 0.00%, max/min/avg concurrent inter-broker partition movements per-broker: 5/5/5.00, 
+ triggeredSelfHealingTaskId: 4b1ed191-50d7-4aa1-806b-b8c5c0d679fe, triggeredTaskReason:
+ Self healing for DISK_FAILURE: {
+	Disk /var/lib/kafka/data-1/kafka-log2 on broker 2 failed at 2024-12-09T09:49:46Z
+ }}
+```
+where,
+* `state` refers to the status of the replica movement(pending/in-progress/finished etc.)
+* `triggeredSelfHealingTaskId` points to the anomalyId
+* `triggeredTaskReason` refers to the reason why self-healing was triggered
+
+### Self-healing configuration in the Strimzi operator
+
+Users can activate self-healing by setting the `self.healing.enabled` property to `true` and specifying a notifier that the self-healing mechanism is going to use in the `spec.cruiseControl.config` section of their `Kafka` resource.
+If the user don't set the notifier property then `NoopNotifer` will be used by default which means all the action on the anomalies would be set to `IGNORE`.
```

Comment:
I would so make clear that the `self.healing.enabled` is not enough by itself then, because you also need to set a proper notifier (instead of the default `NoopNotifier`).

Reply:
But wouldn't that mean that we will then have to create another new notifier? Isn't just specifying enough or should or maybe a warning in the Kafka CR is enough?

Reply:
I guess this discussion is clashing with the other one where we are thinking of setting the Kube events notifier as default, or?

Reply:
Yes, for now I have updated the proposal saying that if no notifier is specified then kube events notifier will be selected as default but i wonder if we should do that?  Shouldn't we just create a `warning/error` in Kafka CR saying that specifying the notifier is must or maybe setting the `SelfHealingNotifer` as Default so that self healing still works and also mentioning in the Kafka CR?

Reply:
The `SelfHealingNotifer` has an implementation of the `alert` method which just log in the CC pod, which is of course not so Kube friendly or at least it means a user should take a look at log. I think in our Kube env, it would be better having something more "cloud-native friendly".